okhttp/src/main/java/okhttp3/internal/ws/CompressionCorrector.kt

```diff
@@ -0,0 +1,42 @@
+@file:JvmName("CompressionCorrector")
+
+package okhttp3.internal.ws
+
+import okio.Buffer
+import okio.ByteString.Companion.decodeHex
+
+private val EMPTY_DEFLATE_BLOCK = "000000ffff".decodeHex()
+private const val LAST_OCTETS_COUNT_TO_REMOVE_AFTER_DEFLATION = 4
+
+private const val OCTETS_TO_ADD_BEFORE_INFLATION = 0x0000ffff
+
+/**
+ * Applies post-deflate transformations described in
+ * [rfc7692#section-7.2.1](https://tools.ietf.org/html/rfc7692#section-7.2.1).
+ *
+ * @param deflated Buffer containing deflated data
+ */
+fun applyPostDeflate(deflated: Buffer) = deflated.apply {
+  if (endsWithEmptyDeflateBlock(this)) {
```

Comment:
The rfc [suggests](https://tools.ietf.org/html/rfc7692#section-7.2.1) the following approach: 

```
   1.  Compress all the octets of the payload of the message using
       DEFLATE.

   2.  If the resulting data does not end with an empty DEFLATE block
       with no compression (the "BTYPE" bits are set to 00), append an
       empty DEFLATE block with no compression to the tail end.

   3.  Remove 4 octets (that are 0x00 0x00 0xff 0xff) from the tail end.
       After this step, the last octet of the compressed data contains
       (possibly part of) the DEFLATE header bits with the "BTYPE" bits
       set to 00.
```

Empty deflate block AFAIK is 5 bytes: `00 00 00 ff ff`. 
So, if the deflated data does not end with it, we first append these 5 bytes, and then remove last 4 of them, leaving the 1st of `00`. 
So, it's guaranteed that deflated messages sent on the wire end with `00`. 
Of inflation, we first append the previously removed 4 bytes of `00 00 ff ff`, so the resulting deflated data is guaranteed to end with empty deflate block `00 00 00 ff ff`. Only after the data is passed to the inflator. 

So, answering your question, if deflated application data ends with `00 00 00 ff ff`, it will be exactly the same before inflation.

In the `else` block here I just add `00` instead of adding `00 00 00 ff ff` and removing `00 00 ff ff`.

Does it make sense?

Reply:
I think I get it. The spec authors recognized that DEFLATE wasted four bytes so they created their own DEFLATE variant to avoid that.

SGtM

Reply:
Exactly.